spark/src/main/java/org/apache/iceberg/spark/data/SparkOrcValueReaders.java

```diff
@@ -195,7 +196,12 @@ public Long nonNullRead(ColumnVector vector, int row) {
     @Override
     public Decimal nonNullRead(ColumnVector vector, int row) {
       HiveDecimalWritable value = ((DecimalColumnVector) vector).vector[row];
-      return new Decimal().set(value.serialize64(value.scale()), value.precision(), value.scale());
+      BigDecimal decimal = new BigDecimal(BigInteger.valueOf(value.serialize64(value.scale())), value.scale());
```

Comment:
`value.serialize64()` will take in an expected scale as a parameter, so I think the only change required to the original code is to pass our expected reader scale into `value.serialize64()` instead of passing `value.scale()`.

Reply:
Sounds great.  The essential purpose here is to construct a `Decimal` with the correct `precision` and `scale` ( instead of the `value.precision()` and `value.scale()`.

Reply:
Oh,  seems it's still incorrect.  Because the `value.serialize64(scale)` is still encoded by `value.precision()`  and `value.scale()`. we use the given `precision` and `scale` to parse this long value,  it will be messed up.  Notice, the value.precision is not equals to `precision`, similar to scale.  

The correct way should be: 
```java
Decimal decimal = new Decimal().set(value.serialize64(value.scale()), value.precision(), value.scale());
decimal.changePrecision(precision, scale);
```

Reply:
I believe `value.serialize64` returns the unscaled long value (since precision <= 18, it always fits in long), I don't think it is tied to any precision. Can you give an example of the case you are referring to?

Reply:
Checked this again,  I wrongly used the `return new Decimal().set(value.serialize64(value.scale()), precision, scale)` to construct the decimal before, which broken the unit tests.  You are right,  the long value is not tied to any precision.  Sorry for the noisy.